### PR TITLE
[LinkerWrapper] Accept some needed lld-link linker arguments for COFF targets

### PR DESCRIPTION
**Pull request by jhuber6**
Summary:
The linker wrapper is a utility used to create offloading programs from
single-source offloading languages such as OpenMP or CUDA. This is done
by embedding device code into the host object, then feeding it into the
linker wrapper which extracts the accelerator object files, links them,
then wraps them in registration code for the target  runtime. This
previously has only worked in Linux / ELF platforms.

This patch attempts to hand Windows / COFF inputs by also accepting COFF
forms of certain linker arguments we use internally. The important
arguments are library search paths, so we can identify libraries which
may contain device code, libraries themselves, and the output name used
for intermediate output.

I am not intimately familiar with the semantics here for the semantics
in how a `lib` file is earched. I am simply treating `foo.lib` as the
GNU equivalent `-l:foo.lib` in the search logic. Similarly, I am
assuming that static libraries will be llvm-ar style libraries. I will
need to investigate the actual deficiencies later, but this should be a
good starting point along with https://github.com/llvm/llvm-project/pull/72697

### CODE REVIEW

**Comment by Meinersbur** [comment]
The command-line argument handling is not related to [PE](https://en.wikipedia.org/wiki/Portable_Executable)/COFF, but to Microsoft's `link.exe` command line interface, for instance [`/libpath:`](https://learn.microsoft.com/en-us/cpp/build/reference/libpath-additional-libpath?view=msvc-170). `/usr/bin/lld-link` is a `link.exe`-compatible interface for lld with an appropriate default triple, like `clang-cl` is for `clang`. IIRC, `lld` choses its command-line interface based on the `argv[0]` name, so should clang-linker-wrapper when passed `--linker-path=/usr/bin/lld-link` instead of `--linker-path=/usr/bin/lld`, but both should be able to generate PE files.

That is, this patch is not necessarily wrong, but the commit message and "// COFF-style linker options." should refer to the command line interface instead.

File: clang/tools/clang-linker-wrapper/LinkerWrapperOpts.td [hunk 1]
```diff
@@ -126,3 +126,8 @@ def version : Flag<["--", "-"], "version">, Flags<[HelpHidden]>, Alias<v>;
 
 def whole_archive : Flag<["--", "-"], "whole-archive">, Flags<[HelpHidden]>;
 def no_whole_archive : Flag<["--", "-"], "no-whole-archive">, Flags<[HelpHidden]>;
+
+// COFF-style linker options.
+def out : Joined<["/", "-", "/?", "-?"], "out:">, Flags<[HelpHidden]>;
```
Comment (Meinersbur) [inline clang/tools/clang-linker-wrapper/LinkerWrapperOpts.td:131]: Why `/?` and `-?` in the permitted option prefix list? Wouldn't this allow e.g. `-?out:`?

**Comment by jhuber6** [comment]
> The command-line argument handling is not related to [PE](https://en.wikipedia.org/wiki/Portable_Executable)/COFF, but to Microsoft's `link.exe` command line interface, for instance [`/libpath:`](https://learn.microsoft.com/en-us/cpp/build/reference/libpath-additional-libpath?view=msvc-170). `/usr/bin/lld-link` is a `link.exe`-compatible interface for lld with an appropriate default triple, like `clang-cl` is for `clang`. IIRC, `lld` choses its command-line interface based on the `argv[0]` name, so should clang-linker-wrapper when passed `--linker-path=/usr/bin/lld-link` instead of `--linker-path=/usr/bin/lld`, but both should be able to generate PE files.
> 
> That is, this patch is not necessarily wrong, but the commit message and "// COFF-style linker options." should refer to the command line interface instead.

I changed the title, realistically I could probably try to separate these more logically, but I think it's easier to just handle them both independently but identically in the logic.

File: clang/tools/clang-linker-wrapper/LinkerWrapperOpts.td (hunk 1 above)
Comment (Meinersbur) [reply clang/tools/clang-linker-wrapper/LinkerWrapperOpts.td:131]: Feels strange to me, but justification is in 4b81e9f8d10214fbb2c8e1a367a60801a5dd5fed

File: clang/tools/clang-linker-wrapper/LinkerWrapperOpts.td [hunk 2]
```diff
@@ -126,3 +126,8 @@ def version : Flag<["--", "-"], "version">, Flags<[HelpHidden]>, Alias<v>;
 
 def whole_archive : Flag<["--", "-"], "whole-archive">, Flags<[HelpHidden]>;
 def no_whole_archive : Flag<["--", "-"], "no-whole-archive">, Flags<[HelpHidden]>;
+
+// COFF-style linker options.
```
Comment (Meinersbur) [inline clang/tools/clang-linker-wrapper/LinkerWrapperOpts.td:130]:
```suggestion
// link.exe-style linker options.
```
Can you change this as well?

**Comment by jhuber6** [comment]
Do COFF / Windows libraries have a special file magic / handling? I may need to update the extraction code to handle those in a later patch.

**Review by Meinersbur — APPROVED** [review]
PE file start with the magic bytes `MZ` (0x4D5A), already recognized by `identify_magic`. Extracting sections can usually done with the [dumpbin.exe](https://learn.microsoft.com/en-us/cpp/build/reference/dumpbin-reference?view=msvc-170) tool. I don't know whether LLVM has means to do it itself as well. The reference for the file format is here: https://learn.microsoft.com/en-us/windows/win32/debug/pe-format

Edit: LLVM has parsing support here: https://github.com/llvm/llvm-project/blob/main/llvm/lib/Object/COFFObjectFile.cpp

**Pull request merged**